lib/descriptor/keyprovider.js

```diff
@@ -0,0 +1,453 @@
+/*!
+ * keyprovider.js - parsed key object for descriptor in bcoin
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const HD = require('../hd/hd');
+const base16 = require('bcrypto/lib/encoding/base16');
+const KeyRing = require('../primitives/keyring');
+const {scriptContext} = require('./common');
+const Network = require('../protocol/network');
+
+/**
+ * Constants
+ */
+
+/**
+ * Derive type for a key in a descriptor
+ * @const {Object}
+ */
+
+const deriveType = {
+  NO: 'NO',
+  UNHARDENED: 'UNHARDENED',
+  HARDENED: 'HARDENED'
+};
+
+/**
+ * KeyProvider
+ * Base class for key object in a descriptor.
+ * @alias module:descriptor.KeyProvider
+ * @property {KeyRing} ring
+ * @property {Number} pos - index of the key in the descriptor
+ * @property {Network} network
+ * @property {String} hardenedMarker - hardened marker for derivation path
+ * default is `h`
+ */
+
+class KeyProvider {
+  /**
+   * @constructor
+   */
+
+  constructor() {
+    this.ring = null;
+    this.pos = 0;
+    this.network = null;
+    this.originInfo = null;
+    this.hardenedMarker = 'h';
+  }
```

Comment:
I think this should be called `AbstractKeyProvider` since it should never be instantiated on its own. The abstract methods should throw errors to enforce its unusability. Finally, I think the common options could be parsed here such as ring, pos, originInfo, etc. For a good example see bcoin's `AbstractBlock`, `Block` and `Headers` classes.

Reply:
Wow! This was a super cool insight. Will implement this idea in descriptor base class as well. I was just thinking should this still be named AbstractKeyProvider? I see that AbstractBlock has no static functions. But I am parsing the descriptor expressions with KeyProvider.fromString(..) function.